### ISSUE TITLE
GetExpectedKernelType 方法的几种写法

### ISSUE THREAD

**Issue by iclementine** [PaddlePaddle/Paddle#14952]
1）PaddlePaddle版本：1.2, commit ID 3628d894f2c5077832fa78f76caa75f979b42a08
    2）CPU/GPU：GPU Driver nvidia 384.81, cuda　8.0.61， cudnn 7.0
    3）系统环境：ubuntu Docker 16.04 LTS
    4）Python版本号: 2.7.12
    5）显存信息
问题描述，在写 operator 的时候， GetExpectedKernelType 方法的写法，这两种有什么区别。

elementwise op 的写法
```c++
  framework::OpKernelType GetExpectedKernelType(
      const framework::ExecutionContext &ctx) const override {
    auto input_data_type = framework::GetDataTypeOfVar(ctx.InputVar("X"));
    return framework::OpKernelType(input_data_type, ctx.GetPlace());
  }
```
因为 OkKernelType 的 constructor 需要的第一个参数类型是 proto::VarType::Type，而 framework::GetDataTypeOfVar 确实返回 proto::VarType::Type。

softmax with cross entropy op 的写法
```c++
 protected:
  framework::OpKernelType GetExpectedKernelType(
      const framework::ExecutionContext& ctx) const override {
    return framework::OpKernelType(
        framework::ToDataType(ctx.Input<Tensor>("Logits")->type()),
        ctx.device_context());
  }
```

因为 `framework/data_type.h` 中声明了两个函数

```c++
extern proto::VarType::Type ToDataType(std::type_index type);
extern std::type_index ToTypeIndex(proto::VarType::Type type);
```
所以可以把 `ctx.Input<Tensor>("Logits")->type()` 返回来的 std::type_index 转换成 `proto::VarType::Type` 这也自然。

但是 nce_op 中的写法就让人觉得奇怪了。
```c++
 protected:
  framework::OpKernelType GetExpectedKernelType(
      const framework::ExecutionContext &ctx) const override {
    return framework::OpKernelType(ctx.Input<Tensor>("Input")->type(),
                                   platform::CPUPlace());
  }
```
`ctx.Input<Tensor>("Input")->type()` 返回的是一个 `std::type_index`, 按理说应该是没有对应的 `framework::OpKernelType` constructor 才对，然而竟然这样写也可以。这是为什么呢？

另外，我照着 softmax with cross entropy op 的写法写的时候，编译的时候会遇到无法将 `std::type_index` 转换成 `proto::VarType::Type` 的错误。请问这是什么原因？后来改成类似 elementwise op 的写法就可以了。

**Comment by Yancey0623** [comment]
应该是之前于老师的一个PR把 tensor::type() 的返回值改成了 `proto::VarType::Type`, 清掉build目录重新编译一下试试呢？

PR: https://github.com/PaddlePaddle/Paddle/pull/14858/files#diff-990fb28c91b41c53e13387d071897f02R141

**Comment by iclementine** [comment]
了解， 谢谢
